server/internal/usecase/interactor/asset.go

```diff
@@ -17,14 +18,18 @@ import (
 )
 
 type Asset struct {
-	repos    *repo.Container
-	gateways *gateway.Container
+	repos     *repo.Container
+	gateways  *gateway.Container
+	eventFunc func(ctx context.Context, wid id.WorkspaceID, t event.Type, a *asset.Asset, op event.Operator) (*event.Event[any], error)
```

Comment:
I defined event func as an element of struct because it is easy to mock this func, otherwise, all use ases depend on it which disturbs us from testing what we truly want.

Reply:
```go
func (i *Asset) createEvent(ctx context.Context, wid id.WorkspaceID, t event.Type, a *asset.Asset, op event.Operator) (*event.Event[any], error) {
	return createEvent(ctx, i.r, i.g, wid, t, a, op)
}

// test

func Test_Asset(t *testing.T) {
	uc := &Asset{
		r: memory.New(),
        g: &gateway.Container{ Task: gatewaymock.Task(...) },
	}

	// test usecases
}
```

Is not good?

Reply:
The reason for doing this is that all use cases that issue events rely on `createEvent`, which returns an error if the workspace does not exist in the DB. Even if you Mock the TaskRunner, you will still get errors caused by other parts of the system. This is not surprising, but from a testing perspective, it is necessary to include the workspace as a seed in all use case tests. To prevent this from affecting all use case tests, etc. when `createEvent` is extended in the future, we have made it possible to replace the function in this way. Of course, this may not be the ideal way to do it, but I tried to do it this way. What do you think?

Reply:
>Also it's good to define function to get asset URLs such:

Implemented [def url resolver](https://github.com/reearth/reearth-cms/pull/362/commits/b910f0d4e905e96e7c541c76c9fd268428375a8f)

Reply:
Ok i see. Go ahead.

Also, asset.URLResolver is better than asset.AssetURLResolver.